contracts/vaults/YearnVaults/base/RibbonVault.sol

```diff
@@ -0,0 +1,815 @@
+        ShareMath.assertUint104(shares);
+
+        Vault.DepositReceipt memory depositReceipt =
+            depositReceipts[msg.sender];
+
+        // This handles the null case when depositReceipt.round = 0
+        // Because we start with round = 1 at `initialize`
+        uint16 currentRound = vaultState.round;
+        require(depositReceipt.round < currentRound, "Round not closed");
+
+        uint128 unredeemedShares =
+            depositReceipt.getSharesFromReceipt(
+                currentRound,
+                roundPricePerShare[depositReceipt.round],
+                vaultParams.decimals
+            );
+
+        shares = isMax ? unredeemedShares : shares;
+        require(shares > 0, "!shares");
+        require(shares <= unredeemedShares, "Exceeds available");
+
+        // This zeroes out any pending amount from depositReceipt
+        depositReceipts[msg.sender].amount = 0;
+        depositReceipts[msg.sender].processed = true;
+        depositReceipts[msg.sender].unredeemedShares = uint128(
+            uint256(unredeemedShares).sub(shares)
+        );
+
+        emit Redeem(msg.sender, shares, depositReceipt.round);
+
+        _transfer(address(this), msg.sender, shares);
+    }
+
+    /************************************************
+     *  VAULT OPERATIONS
+     ***********************************************/
+
+    /*
+     * @notice Helper function that helps to save gas for writing values into the roundPricePerShare map.
+     *         Writing `1` into the map makes subsequent writes warm, reducing the gas from 20k to 5k.
+     *         Having 1 initialized beforehand will not be an issue as long as we round down share calculations to 0.
+     * @param numRounds is the number of rounds to initialize in the map
+     */
+    function initRounds(uint256 numRounds) external nonReentrant {
+        require(numRounds < 52, "numRounds >= 52");
+
+        uint16 _round = vaultState.round;
+        for (uint16 i = 0; i < numRounds; i++) {
+            uint16 index = _round + i;
+            require(index >= _round, "Overflow");
+            require(roundPricePerShare[index] == 0, "Initialized"); // AVOID OVERWRITING ACTUAL VALUES
+            roundPricePerShare[index] = PLACEHOLDER_UINT;
+        }
+    }
+
+    /*
+     * @notice Helper function that performs most administrative tasks
+     * such as setting next option, minting new shares, getting vault fees, etc.
+     * @return newOption is the new option address
+     * @return lockedBalance is the new balance used to calculate next option purchase size or collateral size
+     */
+    function _rollToNextOption() internal returns (address, uint256) {
+        require(block.timestamp >= optionState.nextOptionReadyAt, "Not ready");
+
+        address newOption = optionState.nextOption;
+        require(newOption != address(0), "!nextOption");
+
+        // Wrap entire `asset` balance to `collateralToken` balance
+        _wrapToYieldToken();
```

Comment:
Is it possible to move this to the bottom of the function? We're following the [check-effect-interaction pattern](https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html) so we should do interactions that involve asset transfers last.